core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1367,11 +1392,81 @@ class KafkaApis(val requestChannel: RequestChannel,
     val responseBody = new OffsetsForLeaderEpochResponse(
       replicaManager.getResponseFor(requestInfo)
     )
-    requestChannel.sendResponse(new RequestChannel.Response(request, responseBody))
+    sendResponseExemptThrottle(request, new RequestChannel.Response(request, responseBody))
+  }
+
+  private def handleError(request: RequestChannel.Request, e: Throwable) {
+    if (request.requestObj != null) {
+       def sendResponseCallback(throttleTimeMs: Int) {
+        request.requestObj.handleError(e, requestChannel, request)
+        error("Error when handling request %s".format(request.requestObj), e)
+      }
+      val clientId : String =
+        if (request.requestObj.isInstanceOf[ControlledShutdownRequest])
+          request.requestObj.asInstanceOf[ControlledShutdownRequest].clientId.getOrElse("")
+        else
+          throw new IllegalStateException("Old style requests should only be used for ControlledShutdownRequest")
+      sendResponseMaybeThrottle(request, clientId, sendResponseCallback)
+    } else {
+      def createResponse(throttleTimeMs: Int): AbstractResponse = {
+        val response = request.body[AbstractRequest].getErrorResponse(throttleTimeMs, e)
+
+        /* If request doesn't have a default error response, we just close the connection.
+           For example, when produce request has acks set to 0 */
+        if (response == null)
+          requestChannel.closeConnection(request.processor, request)
+        response
+      }
+      error("Error when handling request %s".format(request.body[AbstractRequest]), e)
+      sendResponseMaybeThrottle(request, createResponse)
```

Comment:
Here, we throttle independent of the exception type. Perhaps, it's better to only engage in throttling if the exception is ClusterAuthorizationException?